### PR TITLE
Adding property grid widget as a package

### PR DESCRIPTION
**Pull request by zjzoltek**
Initial implementation of the property-grid-widget

This PR adds a new project to the ./packages folder named property-grid-react. This project provides a widget controller which can be used to transform a 9-zone widget into one which allows the user access to the property grid

See the README on how to use the package within an application.

### CODE REVIEW

File: packages/property-grid/package.json [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+    "iModel Property Grid",
+    "iModel.js"
+  ],
+  "author": {
+    "name": "Bentley Systems, Inc.",
+    "url": "http://www.bentley.com"
+  },
+  "repository": {
+    "type": "git",
+    "url": "https://github.com/imodeljs/viewer-components-react/tree/master/packages/tree-widget"
```
Comment (aruniverse) [inline packages/property-grid/package.json:29]: update the url

File: packages/property-grid/package.json [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+  },
+  "peerDependencies": {
+    "@bentley/itwin-find-similar": "^3.6.9",
```
Comment (aruniverse) [inline packages/property-grid/package.json:32]: find-similar not oss, also why is this needed in this pkg?

File: packages/property-grid/package.json [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+    "@bentley/imodeljs-frontend": "^2.0.0",
+    "@bentley/imodeljs-i18n": "^2.0.0",
+    "@bentley/imodeljs-markup": "^2.0.0",
+    "@bentley/imodeljs-quantity": "^2.0.0",
+    "@bentley/itwin-client": "^2.0.0",
+    "@bentley/orbitgt-core": "^2.0.0",
+    "@bentley/presentation-common": "^2.0.0",
+    "@bentley/presentation-frontend": "^2.0.0",
+    "@bentley/product-settings-client": "^2.0.0",
+    "@bentley/react-scripts": "3.4.1",
```
Comment (aruniverse) [inline packages/property-grid/package.json:72]: why is react-scripts a dep? should remove that

File: packages/property-grid/package.json [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+    "@types/react": "~16.9.35",
+    "@types/react-dom": "^16.8.0",
+    "@types/react-redux": "^7.0.1",
+    "cpx": "^1.5.0",
+    "rimraf": "^3.0.2",
+    "tslint": "^5.11.0",
+    "typescript": "~3.7.4"
+  },
+  "dependencies": {
+    "@bentley/itwin-find-similar": "^3.6.9",
```
Comment (aruniverse) [inline packages/property-grid/package.json:90]: find-similar again.

File: packages/property-grid/public/locales/en/PropertyGrid.json [hunk 5]
```diff
@@ -0,0 +1,29 @@
+{
+  "widget-label": "Properties Widget",
+  "context-menu": {
+    "find-similar": {
```
Comment (aruniverse) [inline packages/property-grid/public/locales/en/PropertyGrid.json:4]: oh wait, it is being used, then do we need to oss that pkg too? or do we just allow users to add their own components to the properties widget too?

File: packages/property-grid/tsconfig.json [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+  "extends": "./node_modules/@bentley/build-tools/tsconfig-base.json",
+  "compilerOptions": {
+    "target": "es2017",
+    "module": "commonjs",
+    "declarationMap": true,
+    "outDir": "./lib",
+    "skipLibCheck": true,
+    "jsx": "react",
+    "preserveSymlinks": true,
+    "types": [
```
Comment (aruniverse) [inline packages/property-grid/tsconfig.json:11]: can remove the types here

File: packages/property-grid/src/components/PropertyGrid.scss [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+* Licensed under the MIT License. See LICENSE.md in the project root for license terms.
+*--------------------------------------------------------------------------------------------*/
+
+@import "~@bentley/ui-core/lib/ui-core/style/index";
+@import "~@bentley/ui-core/lib/ui-core/text/index";
+
+$light-background-3: #eef0f3;
+$light-background-5: #c7ccd1;
+$light-foreground-body: #000;
+$light-foreground-primary: #008be1;
```
Comment (aruniverse) [inline packages/property-grid/src/components/PropertyGrid.scss:12]: do we need specific values for dark mode?

File: packages/property-grid/src/components/PropertyGrid.scss [hunk 8]
```diff
@@ -0,0 +1,93 @@
+/*---------------------------------------------------------------------------------------------
+* Copyright (c) 2020 Bentley Systems, Incorporated. All rights reserved.
```
Comment (aruniverse) [inline packages/property-grid/src/components/PropertyGrid.scss:2]: need to update the header to use oss one

File: packages/property-grid/package.json [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+  "version": "1.0.0",
+  "description": "Property Grid React",
+  "license": "MIT",
+  "main": "lib/property-grid-react.js",
+  "typings": "lib/property-grid-react",
+  "scripts": {
+    "copy:assets": "cpx \"./src/**/*.scss\" ./lib && cpx \"./public/**/*\" ./lib/public",
+    "pseudolocalize": "betools pseudolocalize --englishDir ./public/locales/en --out ./lib/public/locales/en-PSEUDO",
+    "build": " tsc 1>&2 && npm run copy:assets && npm run pseudolocalize",
+    "clean": "rimraf lib build",
```
Comment (aruniverse) [inline packages/property-grid/package.json:12]: there shouldnt be a build dir

File: packages/property-grid/src/PropertyGridManager.ts [hunk 10]
```diff
@@ -0,0 +1,74 @@
+/*---------------------------------------------------------------------------------------------
+* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
+* Licensed under the MIT License. See LICENSE.md in the project root for license terms.
+*--------------------------------------------------------------------------------------------*/
+/** @module Common */
```
Comment (aruniverse) [inline packages/property-grid/src/PropertyGridManager.ts:5]: can remove this line

File: packages/property-grid/src/PropertyGridManager.ts [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+
+import * as i18next from "i18next";
+import { I18N } from "@bentley/imodeljs-i18n";
+import { UiError, getClassName } from "@bentley/ui-abstract";
+import { FindSimilar } from "@bentley/itwin-find-similar/lib/api/FindSimilar";
```
Comment (aruniverse) [inline packages/property-grid/src/PropertyGridManager.ts:10]: FindSimilar doesnt seem to be used here. But it would also need to be initialized prior to being used in this pkg. Maybe do a lazy init for it in here?

File: packages/property-grid/src/api/WebUtilities.ts [hunk 12]
```diff
@@ -0,0 +1,65 @@
+/*---------------------------------------------------------------------------------------------
+ * Copyright (c) 2020 Bentley Systems, Incorporated. All rights reserved.
```
Comment (aruniverse) [inline packages/property-grid/src/api/WebUtilities.ts:2]: need to update the header

Comment (aruniverse) [reply packages/property-grid/src/api/WebUtilities.ts:2]: also this should probably be moved to a helper pkg cause i think i saw this exact file in tree-widget

File: packages/property-grid/src/api/PropertiesGridSelectionContext.ts [hunk 13]
```diff
@@ -0,0 +1,133 @@
...
+  private static _selectionEvent:
+    | BeEvent<(ids: Set<string>) => void>
+    | undefined;
+  private static _suspended: boolean = false;
+
+  /**
+   * Initialize the properties grid selection context to provide automatic opening and closing of the property panel
+   * @param _selectionEvent Event to use for selection set change listening
+   * @param __propertyPanelId App given Id for the property panel widget
+   * @param widgetIds Ids of widgets that should take precendence and avoid the panel from opening at all if open (e.g. Find Similar Widget)
```
Comment (aruniverse) [inline packages/property-grid/src/api/PropertiesGridSelectionContext.ts:28]: update comment here depending on what we do with find-similar

File: packages/property-grid/src/components/PropertyGrid.tsx [hunk 14]
```diff
@@ -0,0 +1,537 @@
...
+          this.props.iModelConnection!,
+          nextProps.rulesetId,
+        );
+        this._dataProvider.onDataChanged.addListener(this._dataChangedHandler);
+      }
+    }
+  }
+
+  /** Used in tests */
+  private debugConsolePropertyData(data: PropertyData) {
```
Comment (aruniverse) [inline packages/property-grid/src/components/PropertyGrid.tsx:141]: handy, but there are no test for this pkg yet

File: packages/property-grid/src/components/PropertyGrid.tsx [hunk 15]
```diff
@@ -0,0 +1,537 @@
...
+
+  private _shareActionButtonRenderer: ActionButtonRenderer = (
+    props: ActionButtonRendererProps
+  ) => {
+    const shared =
+      this.state.sharedFavorites !== undefined &&
+      this.state.sharedFavorites?.findIndex(
+        (fav: string) => props.property.property.name === fav
+      ) >= 0;
+    // return <div>{shared && <span className="icon icon-share" style="padding-right: 5px;"></span>}</div>;
```
Comment (aruniverse) [inline packages/property-grid/src/components/PropertyGrid.tsx:324]: can remove the commented out code

**Comment by aruniverse** [comment]
Also dont forget to update the code-owners file

File: packages/property-grid/src/api/WebUtilities.ts (hunk 12 above)
Comment (zjzoltek) [reply packages/property-grid/src/api/WebUtilities.ts:2]: I'll take care of this in a separate PR

File: packages/property-grid/src/components/PropertyGrid.scss (hunk 7 above)
Comment (zjzoltek) [reply packages/property-grid/src/components/PropertyGrid.scss]: changed these to buic colors

File: packages/property-grid/src/api/PropertiesGridSelectionContext.ts (hunk 13 above)
Comment (zjzoltek) [reply packages/property-grid/src/api/PropertiesGridSelectionContext.ts]: decided to go ahead and remove find similar and allow users to add their own context menu options (therefore i adjusted this comment)

**Comment by zjzoltek** [comment]
@aruniverse decided to go ahead and remove all find similar usages in favor of allowing users to supply their own additional context menu options

File: packages/property-grid/src/components/PropertyGrid.scss [hunk 16]
```diff
@@ -0,0 +1,88 @@
...
+  font-size: $uicore-font-size-leading;
+  padding: $uicore-xs 0;
+}
+
+.property-grid-property-panel-class {
+  @include uicore-text;
+  font-size: $uicore-font-size;
+}
+
+.property-grid-navigator-property-grid {
```
Comment (aruniverse) [inline packages/property-grid/src/components/PropertyGrid.scss:20]: id remove navigator in the class name, and maybe clean up some of the others as well

File: packages/property-grid/src/components/PropertyGrid.tsx [hunk 17]
```diff
@@ -0,0 +1,530 @@
...
+            "context-menu.add-favorite.description"
+          ),
+          label: PropertyGridManager.translate(
+            "context-menu.add-favorite.label"
+          ),
+        });
+      }
+    }
+
+    // Find Similar Phase 2 adds a 'copy text' context menu option
```
Comment (aruniverse) [inline packages/property-grid/src/components/PropertyGrid.tsx:416]: cleanup anywhere than mentions find-similar

File: README.md [hunk 18]
```diff
@@ -8,11 +8,11 @@ With the repository containing many different packages related to iModel.js, eac
...
-| Name                         | Folder                 |
-| ---------------------------- | ---------------------- |
-| @bentley/imodel-select-react | packages/imodel-select |
-| @bentley/tree-widget-react   | packages/tree-widget   |
-|                              |                        |
+| Name                         | Folder                       |
+| ---------------------------- | ---------------------------- |
+| @bentley/imodel-select-react | packages/imodel-select       |
+| @bentley/tree-widget-react   | packages/tree-widget         |
+| @bentley/property-grid-react | packages/property-grid-react |
```
Comment (aruniverse) [inline README.md:15]: folder should be packages/property-grid

File: packages/property-grid/package.json [hunk 19]
```diff
@@ -0,0 +1,97 @@
+{
+  "name": "@bentley/property-grid-react",
+  "version": "1.0.0",
+  "description": "Property Grid React",
```
Comment (calebmshafer) [inline packages/property-grid/package.json:4]: May want to add something about iModel.js in the description.

File: packages/property-grid/package.json [hunk 20]
```diff
@@ -0,0 +1,97 @@
...
+    "@types/react-redux": "^7.0.1",
+    "cpx": "^1.5.0",
+    "rimraf": "^3.0.2",
+    "tslint": "^5.11.0",
+    "typescript": "~3.7.4"
+  },
+  "dependencies": {
+    "classnames": "^2.2.5",
+    "i18next": "^10.2.2",
+    "react": "^16.8.0",
```
Comment (calebmshafer) [inline packages/property-grid/package.json:92]: The `react` and `react-dom` dependencies should probably be peers.  That's how most of the other shared packages are setup.

File: packages/property-grid/package.json [hunk 21]
```diff
@@ -0,0 +1,97 @@
...
+    "@bentley/product-settings-client": "^2.0.0",
+    "@bentley/ui-abstract": "^2.0.0",
+    "@bentley/ui-components": "^2.0.0",
+    "@bentley/ui-core": "^2.0.0",
+    "@bentley/ui-framework": "^2.0.0",
+    "@bentley/ui-ninezone": "^2.0.0",
+    "@bentley/webgl-compatibility": "^2.0.0"
+  },
+  "devDependencies": {
+    "@bentley/bentleyjs-core": "^2.0.0",
```
Comment (aruniverse) [inline packages/property-grid/package.json:56]: core bentley imjs dev deps shouldnt be using `^` and should be pinned down

File: packages/property-grid/package.json [hunk 22]
```diff
@@ -0,0 +1,97 @@
...
+  "description": "This package offers a React widget which provides access to the property grid within an iModel. Designed for use in an iModeljs 9-zone application",
+  "license": "MIT",
+  "main": "lib/property-grid-react.js",
+  "typings": "lib/property-grid-react",
+  "scripts": {
+    "copy:assets": "cpx \"./src/**/*.scss\" ./lib && cpx \"./public/**/*\" ./lib/public",
+    "pseudolocalize": "betools pseudolocalize --englishDir ./public/locales/en --out ./lib/public/locales/en-PSEUDO",
+    "build": " tsc 1>&2 && npm run copy:assets && npm run pseudolocalize",
+    "clean": "rimraf lib",
+    "lint": "tslint -p . 1>&2",
```
Comment (calebmshafer) [inline packages/property-grid/package.json:13]: This currently doesn't do anything.  Did you mean to add a `tslint.json` or config to the `package.json`?

File: packages/property-grid/package.json (hunk 21 above)
Comment (aruniverse) [reply packages/property-grid/package.json]: sorry, just had a discussion with @calebmshafer and jason, and we're going to use `^` for core imjs deps as well now. please ignore my earlier comment, and revert that change

File: packages/property-grid/package.json (hunk 22 above)
Comment (calebmshafer) [reply packages/property-grid/package.json:13]: Going to resolve this for now.  We need to revisit listing in all packages in the repo.

Comment (zjzoltek) [reply packages/property-grid/package.json:13]: added a tslint.json. Currently it just extends the tslint configuration found in `@bentley/build-tools/tslint-docs.json` and disables one rule. LMK if you have any recommendations for further modification here

**Comment by aruniverse** [comment]
fyi @zjzoltek bad merge with the main readme, look at archanas commit on master before your changes. can you fix it when you create a new pr integrating spencers changes here?
